core/cache/redisBackend.go

```diff
@@ -0,0 +1,378 @@
+/**
+ * @TODO:
+ * - write documentation
+ */
+
+package cache
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"regexp"
+	"runtime"
+	"time"
+
+	"flamingo.me/flamingo/v3/framework/flamingo"
+	"github.com/gomodule/redigo/redis"
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+)
+
+type (
+	// RedisBackend instance representation
+	RedisBackend struct {
+		BackendHitCounter
+		Pool             *redis.Pool
+		logger           flamingo.Logger
+		writeLockEnabled bool
+	}
+
+	// RedisCacheEntryMeta representation
+	RedisCacheEntryMeta struct {
+		Lifetime, Gracetime time.Duration
+	}
+
+	// RedisCacheEntry representation
+	RedisCacheEntry struct {
+		Meta RedisCacheEntryMeta
+		Data interface{}
+	}
+
+	// RedisBackendOptions ...
+	RedisBackendOptions struct {
+		Network string
+		// Redis Host (default 127.0.0.1)
+		Host string
+		// Redis Port (default 6379)
+		Port string
+		// Redis Database Number to use (default 0)
+		Db int
+		// Passwort for the rdis connection. empty string for none (default empty)
+		Password string
+		// Maximum number of idle redis connections in the pool. (default 8)
+		MaxIdle int
+		// Timout to close idle connections (default 5m)
+		IdleTimeout time.Duration
+		// WriteLock feature switch
+		WriteLock bool
+	}
+)
+
+const (
+	lockPrefix  = "lock:"
+	tagPrefix   = "tag:"
+	valuePrefix = "value:"
+)
+
+var (
+	redisKeyRegex = regexp.MustCompile(`[^a-zA-Z0-9]`)
+)
+
+func redisConnector(network, address, password string) (redis.Conn, error) {
+	c, err := redis.Dial(network, address)
+	if err != nil {
+		return nil, err
+	}
+	if password != "" {
+		if _, err := c.Do("AUTH", password); err != nil {
+			c.Close()
+			return nil, err
+		}
+	}
+	return c, err
+}
+
+func finalizer(b *RedisBackend) {
+	b.Close()
+}
+
+// DefaultRedisBackendOptions gets the default options for redis backend
+func DefaultRedisBackendOptions() RedisBackendOptions {
+	return RedisBackendOptions{
+		Network:     "tcp",
+		Host:        "127.0.0.1",
+		Port:        "6379",
+		Db:          0,
+		Password:    "",
+		MaxIdle:     8,
+		IdleTimeout: time.Minute * 30,
+		WriteLock:   false,
+	}
+}
+
+// NewRedisBackend creates an redis cache backend
+func NewRedisBackend(options RedisBackendOptions, logger flamingo.Logger) Backend {
+	err := mergo.Merge(&options, DefaultRedisBackendOptions())
+	if err != nil {
+		logger.WithField("category", "redisBackend").Error(fmt.Sprintf("Error merging options: %v", err))
+	}
+
+	b := &RedisBackend{
+		Pool: &redis.Pool{
+			MaxIdle:     options.MaxIdle,
+			IdleTimeout: options.IdleTimeout,
+			TestOnBorrow: func(c redis.Conn, t time.Time) error {
+				_, err := c.Do("PING")
+				return err
+			},
+			Dial: func() (redis.Conn, error) {
+				return redisConnector(options.Network, fmt.Sprintf("%v:%v", options.Host, options.Port), options.Password)
+			},
+		},
+		logger: logger,
+	}
+	b.withBackendType("redis")
+
+	runtime.SetFinalizer(b, finalizer) // close all connections on destruction
```

Comment:
just learned about that :-) - sounds useful

Reply:
Finalizers are not destructors or anything! They might not be called, from docs:

The finalizer is scheduled to run at some arbitrary time after the program can no longer reach the object to which obj points. There is no guarantee that finalizers will run before a program exits, so typically they are useful only for releasing non-memory resources associated with an object during a long-running program

Reply:
Finalizers are not destructors or anything! They might not be called, from docs:

The finalizer is scheduled to run at some arbitrary time after the program can no longer reach the object to which obj points. There is no guarantee that finalizers will run before a program exits, so typically they are useful only for releasing non-memory resources associated with an object during a long-running program

Reply:
How about using the `flamingo.ShutdownEvent` instead?

Reply:
It was mere an informational info: closing the connection is only necessary when the flamingo stays connected, and/or to finish/flush data.
If the process ends, the connection is closed (by the operating system). We can use flamingo.ShutdownEvent to be sure that the connection is closed, or finalizers to "probably close" it if, by any chance, the cache backend is garbage collected (which actually shouldn't happen?).

So, I'm not against finalizers here, it's mere an info that this is not a destructor as e.g. in java/c++.

Reply:
ah ok. i only added the finalizer here, to ensure the connections are closed.
the reason was:
After adding the docker-tests, the docker containers were not destroyed, because of open connections. after adding the finalizer the connections there closed correctly and the containers are gone ofter the tests are done.

as you described, in production it does not matter. But currently i don't have an other solution, expect adding and exported "Close"-method to the RedisCacheBackend. But exporting an Method only for teardown an test - i am not sure ;)